### ISSUE TITLE
[rllib] pytorch PPO running slower than it was in 0.8.6

### ISSUE THREAD

**Issue by pmacalpine** [ray-project/ray#10003]
### What is the problem?

I've seen a noticeable speed decrease and longer running times when using pytorch PPO with ray 0.9.0.dev0 as compared to ray 0.8.6.  Running times on simple environments like cartpole and pong are somewhere around 10-25% slower.  Also, when using a seed value for cartpole and no GPU, I get different values for pytorch PPO with ray 0.9.0.dev0 and 0.8.6.  I get the same value with both versions of ray with tf which also matches the value with torch and ray 0.8.6.  I'm not sure if it's expected or intended for values to change with torch PPO between the versions of ray when using the same seed.  The slowdown with PPO torch from ray 0.8.6, while not massive, seems problematic and might become worse with more complex models (I will look into testing this with an impala CNN).

*Ray version and other system information (Python version, TensorFlow version, OS):*
ray: 0.9.0.dev0 (latest wheels installed on 8/8/20), 0.8.6
python: 3.6.10
torch: 1.3.1
tensorflow: 1.15.0            
tensorflow-gpu: 2.2.0 
OS: Ubuntu 18.04
Machine: Azure NC6

### Reproduction (REQUIRED)
Please provide a script that can be run to reproduce the issue. The script should have **no external library dependencies** (i.e., use fake or mock data / environments):

Below are slightly modified versions of carpole-ppo and pong-ppo yaml files from tuned_examples/ppo/ with each set to run for 100k total steps with a seed value.

When running cartpole-ppo for ray 0.8.6 with torch and also tf ray 0.8.6/0.9.0.dev0 running time is ~100s and they all return the same reward at the end of 182.01.  With ray 0.9.0.dev0 and torch running time is ~110s and has a reward of 187.25 at the end.

When running pong-ppo for ray 0.8.6 with torch and also tf ray 0.8.6/0.9.0.dev0 running time is ~73s.  With ray 0.9.0.dev0 and torch running time is ~94s

To run the scripts
```
rllib train -f cartpole-ppo.yaml
rllib train -f pong-ppo.yaml
```

```
# cartpole-ppo.yaml
cartpole-ppo:
    env: CartPole-v0
    run: PPO
    stop:
        #episode_reward_mean: 150
        timesteps_total: 100000
    config:
        # Works for both torch and tf.
        framework: torch
        gamma: 0.99
        lr: 0.0003
        num_workers: 1
        observation_filter: MeanStdFilter
        num_sgd_iter: 6
        vf_share_layers: true
        vf_loss_coeff: 0.01
        model:
          fcnet_hiddens: [32]
          fcnet_activation: linear
        seed: 1
```

```
# pong-ppo.yaml
pong-ppo:
    env: PongNoFrameskip-v4
    run: PPO
    stop:
        timesteps_total: 100000
    config:
        # Works for both torch and tf.
        framework: torch
        lambda: 0.95
        kl_coeff: 0.5
        clip_rewards: True
        clip_param: 0.1
        vf_clip_param: 10.0
        entropy_coeff: 0.01
        train_batch_size: 5000
        rollout_fragment_length: 20
        sgd_minibatch_size: 500
        num_sgd_iter: 10
        num_workers: 5 #32
        num_envs_per_worker: 5
        batch_mode: truncate_episodes
        observation_filter: NoFilter
        vf_share_layers: true
        num_gpus: 1
        model:
            dim: 42
        seed: 1
```

If we cannot run your script, we cannot fix your issue.

- [x] I have verified my script runs in a clean environment and reproduces the issue.
- [x] I have verified the issue also occurs with the [latest wheels](https://docs.ray.io/en/latest/installation.html).

**Comment by ericl** [comment]
0.8.6
```
  sampler_perf:
    mean_env_wait_ms: 7.362898637292983
    mean_inference_ms: 3.4891626212778464
    mean_processing_ms: 1.3470325874776283
  time_since_restore: 82.77108430862427
  time_this_iter_s: 15.532827615737915
  time_total_s: 82.77108430862427
  timers:
    learn_throughput: 381.842
    learn_time_ms: 13094.408
    sample_throughput: 1450.34
    sample_time_ms: 3447.467
    update_time_ms: 6.262
```

nightly:
```
  sampler_perf:
    mean_env_wait_ms: 7.178161623474959
    mean_inference_ms: 3.4624236650978872
    mean_processing_ms: 1.104088936852159
  time_since_restore: 82.81693243980408
  time_this_iter_s: 15.645047903060913
  time_total_s: 82.81693243980408
  timers:
    learn_throughput: 378.906
    learn_time_ms: 13195.872
    sample_throughput: 1490.027
    sample_time_ms: 3355.645
    update_time_ms: 5.549
```

I gave this a try and the perf stats (train -v) seem pretty identical to me. Do you see any particular stat that got slower on your setup?

**Comment by pmacalpine** [comment]
The main thing I notice is that learn_time_ms is almost 2.5X more for the nightly build vs 0.8.6.

For 0.8.6
```
  perf:
    cpu_util_percent: 24.13333333333333
    gpu_util_percent0: 0.0
    ram_util_percent: 3.6
    vram_util_percent0: 0.0
  sampler_perf:
    mean_env_wait_ms: 0.0632865910180354
    mean_inference_ms: 0.6566249968717238
    mean_processing_ms: 0.1823716537664795
  time_since_restore: 99.77181482315063
  time_this_iter_s: 3.874606132507324
  time_total_s: 99.77181482315063
  timers:
    learn_throughput: 16392.497
    learn_time_ms: 244.014
    load_throughput: 1831014.1
    load_time_ms: 2.185
    sample_throughput: 1097.026
    sample_time_ms: 3646.222
    update_time_ms: 1.726
  
== Status ==
Memory usage on this node: 4.0/110.0 GiB
Using FIFO scheduling algorithm.
Resources requested: 0/6 CPUs, 0/1 GPUs, 0.0/66.26 GiB heap, 0.0/22.31 GiB objects
Result logdir: /home/patmac/ray_results/cartpole-ppo
Number of trials: 1 (1 TERMINATED)
+-----------------------------+------------+-------+--------+------------------+--------+----------+
| Trial name                  | status     | loc   |   iter |   total time (s) |     ts |   reward |
|-----------------------------+------------+-------+--------+------------------+--------+----------|
| PPO_CartPole-v0_cb089_00000 | TERMINATED |       |     25 |          99.7718 | 100000 |   182.01 |
+-----------------------------+------------+-------+--------+------------------+--------+----------+
```

For nightly:
```
  perf:
    cpu_util_percent: 21.333333333333332
    gpu_util_percent0: 0.0
    ram_util_percent: 3.6
    vram_util_percent0: 0.0
  sampler_perf:
    mean_env_wait_ms: 0.0670671730926781
    mean_inference_ms: 0.6984569164553384
    mean_processing_ms: 0.1591172438601967
  time_since_restore: 110.85516619682312
  time_this_iter_s: 4.29492974281311
  time_total_s: 110.85516619682312
  timers:
    learn_throughput: 6801.903
    learn_time_ms: 588.071
    sample_throughput: 1070.694
    sample_time_ms: 3735.896
    update_time_ms: 1.373
  
== Status ==
Memory usage on this node: 3.9/110.0 GiB
Using FIFO scheduling algorithm.
Resources requested: 0/6 CPUs, 0/1 GPUs, 0.0/66.26 GiB heap, 0.0/22.31 GiB objects (0/1.0 GPUType:P100)
Result logdir: /home/patmac/ray_results/cartpole-ppo
Number of trials: 1 (1 TERMINATED)
+-----------------------------+------------+-------+--------+------------------+--------+----------+
| Trial name                  | status     | loc   |   iter |   total time (s) |     ts |   reward |
|-----------------------------+------------+-------+--------+------------------+--------+----------|
| PPO_CartPole-v0_7a1d6_00000 | TERMINATED |       |     25 |          110.855 | 100000 |   187.25 |
+-----------------------------+------------+-------+--------+------------------+--------+----------+
```

**Comment by ericl** [comment]
Hmm weird I get about the same learn throughput (`learn_throughput: 6593.48` vs `learn_throughput: 6220.133`) on both, whereas you see it much faster with the previous Ray version.

I'm using torch 1.4.0 / python 3.6 here.

**Comment by pmacalpine** [comment]
I see the same results with  torch 1.4.0.  Is it also odd that I get a different reward result with the same seed when using 0.8.6 vs 0.8.7/nightly?

I also see the following tensorflow warning message printed with 0.8.7/nightly, but not 0.8.6, even though I'm not using tf in case that's a concern: 
```
/home/patmac/.conda/envs/dilbert/lib/python3.6/site-packages/tensorflow/python/compat/v2_compat.py:96: disable_resource_variables (from tensorflow.python.ops.variable_scope) is deprecated and will be removed in a future version.
Instructions for updating:
non-resource variables are not supported in the long term
```
